optimum/gptq/quantizer.py

```diff
@@ -0,0 +1,629 @@
+# coding=utf-8
+# Copyright 2023 HuggingFace Inc. team and GPTQ and AutoGPTQ authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import copy
+import json
+import os
+from logging import getLogger
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import torch
+from torch import nn
+from tqdm.auto import tqdm
+from transformers import AutoTokenizer
+from transformers.pytorch_utils import Conv1D
+
+from ..utils import is_accelerate_available, is_auto_gptq_available
+from ..utils.modeling_utils import recurse_getattr
+from .constants import GPTQ_CONFIG
+from .data import get_dataset, prepare_dataset
+from .utils import get_block_name_with_pattern, get_device, get_layers, get_preceding_modules, get_seqlen
+
+
+if is_accelerate_available():
+    from accelerate import Accelerator, cpu_offload_with_hook, load_checkpoint_and_dispatch
+    from accelerate.hooks import remove_hook_from_module
+
+if is_auto_gptq_available():
+    from auto_gptq.nn_modules.qlinear import GeneralQuantLinear
+    from auto_gptq.quantization import GPTQ
+    from auto_gptq.utils.import_utils import dynamically_import_QuantLinear
+
+logger = getLogger(__name__)
+
+
+class GPTQQuantizer(object):
+    r"""
+    A simple API for GPTQ Quantization
+    """
+
+    def __init__(
+        self,
+        bits: int,
+        dataset: Optional[Union[List[str], str]] = None,
+        group_size: int = 128,
+        damp_percent: float = 0.01,
+        desc_act: bool = True,
+        sym: bool = True,
+        true_sequential: bool = True,
+        pack_sequentially: bool = False,
+        use_cuda_fp16: bool = False,
+        model_seqlen: Optional[int] = None,
+        block_name_to_quantize: Optional[str] = None,
+        module_name_preceding_first_block: Optional[List[str]] = None,
+        batch_size: int = 1,
+        pad_token_id: Optional[int] = None,
+        *args,
+        **kwargs,
+    ):
+        """
+        Args:
+            bits (`int`):
+                The number of bits to quantize to, supported numbers are (2, 3, 4, 8).
+            dataset (`Union[List[str],str]`, defaults to None):
```

Comment:
We don't need the dataset when loading quantized weights. I decided to put `None`  because the repo from `TheBloke` don't have a dataset field. This way, we don't need to pass a hardcoded dataset to load these models.

Reply:
Thank you. In my opinion this implementation is not the most elegant one, there is in my opinion two distinct tasks:
* Quantizing a model that is in fp16, for which you need to know all of the `damp_percent`, `dataset`, `batch_size`, `true_sequential`, etc. and saving it to disk.
* Loading a model *that is already quantized*, for which you only need to have the knowledge of bits, g_idx, qweights, qzeros, scales, group_size and possibly sym. Having the knowledge of which layers were replaced is known from the state_dict keys.

Here it feels like the `GPTQQuantizer` mixes up the two tasks, by hacking the `__init__` signature with defaults that should not really be there.

It is fine to keep as is, but I think a simpler design dissociating the two could have been possible. Let me know your thoughts!

Reply:
Totally agree with you. At first, I also wanted to separe these two tasks. However, it was not compatible with Transformers integration if we wanted to quantize a model or load an already quantized model (must have a `quantization_config`) with only the `from_pretrained()`method. I had to put all the arguments inside the `GPTQQuantizer`.